### ISSUE TITLE
Sudo permissions limitations

### ISSUE THREAD

**Issue by p-rintz** [Ravinou/borgwarehouse#5]
Hi there,

I havent looked too far into how it could be better yet, but I'm creating this issue so I don't forget about it again.

And maybe you already have an idea how to make it better.

The issue is that currently in the docs you recommend limiting the borgwarehouse users' sudo permissions to some commands. This is, of course, great. However, you allow things like "/usr/bin/bash", which makes it pretty useless in the current iteration, in my opinion. If you have sudo access to bash, you have sudo access to everything.

Maybe it would be best to make it more fine-grained in allowing access only to the specific scripts that you need to start via bash, however then you'd also need to keep in mind that the scripts you want to start this way shouldnt be editable by the user, otherwise it again defeats the purpose.

I'll come back to this on the weekend probably with a PR, if you haven't already fixed it. :smile:

**Comment by Ravinou** [comment]
Hey Philipp 👋

The sudo permissions I allow are strictly defined according to the system scripts I have created. They are located in the `helper/shell` folder. All commands allowed in `sudo` for the user `borgwarehouse` are necessary to run these scripts, **no more and no less**.

Bash (as well as all other commands allowed in `sudo` unless I'm mistaken) is necessary to run these scripts, if you remove this permission, there will be a password request for elevation of privilege and the use of system functions will not be possible.

If you know a solution to remove the authorization granted for the use of bash I am interested. Are you sure it is possible without modifying scripts ?

**Comment by p-rintz** [comment]
Issue is, currently, it gives more permissions than needed. See the following example:

```
borgwarehouse@backup:~$ sudo apt update

We trust you have received the usual lecture from the local System
Administrator. It usually boils down to these three things:

    #1) Respect the privacy of others.
    #2) Think before you type.
    #3) With great power comes great responsibility.

[sudo] password for borgwarehouse: 
sudo: a password is required
borgwarehouse@backup:~$ sudo bash -c "apt update"
Hit:1 http://mirror.terrahost.no/debian bullseye InRelease
Hit:2 http://mirror.terrahost.no/debian bullseye-updates InRelease                                
Hit:3 http://security.debian.org/debian-security bullseye-security InRelease                      
Hit:4 https://deb.nodesource.com/node_18.x bullseye InRelease                        
Reading package lists... Done                                  
Building dependency tree... Done
Reading state information... Done
All packages are up to date.
```

One solution would be to make sure only the specific scripts needed for borgwarehouse gets allowed via sudo.

ie.
`borgwarehouse ALL=NOPASSWD: /usr/bin/somescript -c with -d some -d parameters-that-are-needed`

That way, only exactly what is needed can be run. But you'd also need to be careful that the scripts cant be edited. Otherwise you again give the power to run everything, since you could simply edit the script and then run the script with the allowed command again to run something totally different.

You can also use regex and something like this for example if you need things to be more dynamic:
`john	ALL = /usr/bin/passwd ^[a-zA-Z0-9_]+$,  !/usr/bin/passwd root`

Documentation on this: https://www.sudo.ws/docs/man/sudoers.man/#Wildcards

For specific ideas how to edit the sudoers permissions, Ill have to look into what the scripts to in detail. Ill do that in the coming days.

**Comment by Ravinou** [comment]
Thank you. Indeed it must be improved. Limiting the use of scripts without the possibility to edit them seems to me a good idea. I would have to find projects with similar cases, there must be some.

I expect you to return in the next few days. Thanks for your help 🙏

**Comment by p-rintz** [comment]
Ill keep this issue open until you have had time to change the documentation. (or you can close it yourself then if you want to)

**Comment by Ravinou** [comment]
Your changes to the sudoers file are interesting, I'm thinking of putting them in the FAQ as regexes are not supported in the latest Debian. Unless you see an alternative solution?

However, by removing bash already, it is now even more limited!

**Comment by p-rintz** [comment]
No, that was my idea for the changes in the PR. To use the rules I mentioned in the PR.

Giving sudo rights to bash directly shouldn't really ever be needed.
As you can see it was possible to do everything you did before in different ways.

Especially now, that the repositories are readable by the borgwarehouse user, there is less reason for elevated permissions.

**Comment by Ravinou** [comment]
Okay, with your PR, the sudoers file can be changed from : 

`borgwarehouse ALL=(ALL) NOPASSWD: /usr/sbin/useradd,/usr/bin/mkdir,/usr/bin/touch,/usr/bin/chmod,/usr/bin/chown,/usr/bin/bash,/usr/bin/jc,/usr/bin/jq,/usr/bin/sed,/usr/bin/grep,/usr/bin/stat,/usr/bin/borg,/usr/bin/echo,/usr/sbin/userdel
`
to : 
```
borgwarehouse ALL=(ALL) NOPASSWD: /usr/sbin/useradd,/usr/bin/mkdir,/usr/bin/touch,/usr/bin/chmod,/usr/bin/chown,/usr/bin/sed,/usr/sbin/userdel,/usr/bin/tee
```

**Comment by p-rintz** [comment]
That doesnt really solve the security issues, sadly.

Just removing bash from the file solves one of it.

But even with the others, you can do pretty bad stuff.

Like:

`chmod 666 /etc/shadow`


That is why I made the sudoers file more fine-grained:

```
borgwarehouse ALL=(ALL) NOPASSWD: /usr/sbin/userdel -rf [[\:xdigit\:]]*
borgwarehouse ALL=(ALL) NOPASSWD: /usr/sbin/useradd -d /var/borgwarehouse/[[\:xdigit\:]]* -s /bin/bash -m [[\:xdigit\:]]*
borgwarehouse ALL=(ALL) NOPASSWD: /usr/bin/mkdir -p /var/borgwarehouse/[[\:xdigit\:]]*/repos/repo[[\:digit\:]]*, /usr/bin/mkdir -p /var/borgwarehouse/[[\:xdigit\:]]*/.ssh
borgwarehouse ALL=(ALL) NOPASSWD: /usr/bin/touch /var/borgwarehouse/[[\:xdigit\:]]*/.ssh/authorized_keys
borgwarehouse ALL=(ALL) NOPASSWD: /usr/bin/chmod -R 750 /var/borgwarehouse/[[\:xdigit\:]]*, /usr/bin/chmod 600 /var/borgwarehouse/[[\:xdigit\:]]*/.ssh/authorized_keys
borgwarehouse ALL=(ALL) NOPASSWD: /usr/bin/chown -R [[\:xdigit\:]]*\:borgwarehouse /var/borgwarehouse/[[\:xdigit\:]]*
borgwarehouse ALL=(ALL) NOPASSWD: /usr/bin/tee /var/borgwarehouse/[[\:xdigit\:]]*/.ssh/authorized_keys
borgwarehouse ALL=(ALL) NOPASSWD: /usr/bin/sed -ri s|*|g /var/borgwarehouse/[[\:xdigit\:]]*/.ssh/authorized_keys
```

**Comment by Ravinou** [comment]
But only work with with sudo version > 1.9.10, is it right ?

**Comment by p-rintz** [comment]
No. That above isnt regex. They are Unix-style wildcards.

Documentation is here: https://facelessuser.github.io/wcmatch/fnmatch/

**Comment by Ravinou** [comment]
ah ok that's what I didn't understand in your answer actually.

**Comment by p-rintz** [comment]
Ah, yeah. Guess I wrote it confusingly.

Here is the difference for one rule as example:

Only works on sudo >1.9.10:
`borgwarehouse ALL=(ALL) NOPASSWD: /usr/sbin/useradd -d /var/borgwarehouse/[a-f0-9]{8} -s /bin/bash -m [a-f0-9]{8}$`

Works on everything, but is less fine-grained:
`borgwarehouse ALL=(ALL) NOPASSWD: /usr/sbin/useradd -d /var/borgwarehouse/[[\:xdigit\:]]* -s /bin/bash -m [[\:xdigit\:]]*`

**Comment by Ravinou** [comment]
Okay, I understand better now. With regexes, you can rely on the fact that I generate strings of 8 characters no more and no less. So it adds another layer of security but it is only possible with sudo v1.9.10.

So, in the documentation, I can replace sudoers with : 
```
borgwarehouse ALL=(ALL) NOPASSWD: /usr/sbin/userdel -rf [[\:xdigit\:]]*
borgwarehouse ALL=(ALL) NOPASSWD: /usr/sbin/useradd -d /var/borgwarehouse/[[\:xdigit\:]]* -s /bin/bash -m [[\:xdigit\:]]*
borgwarehouse ALL=(ALL) NOPASSWD: /usr/bin/mkdir -p /var/borgwarehouse/[[\:xdigit\:]]*/repos/repo[[\:digit\:]]*, /usr/bin/mkdir -p /var/borgwarehouse/[[\:xdigit\:]]*/.ssh
borgwarehouse ALL=(ALL) NOPASSWD: /usr/bin/touch /var/borgwarehouse/[[\:xdigit\:]]*/.ssh/authorized_keys
borgwarehouse ALL=(ALL) NOPASSWD: /usr/bin/chmod -R 750 /var/borgwarehouse/[[\:xdigit\:]]*, /usr/bin/chmod 600 /var/borgwarehouse/[[\:xdigit\:]]*/.ssh/authorized_keys
borgwarehouse ALL=(ALL) NOPASSWD: /usr/bin/chown -R [[\:xdigit\:]]*\:borgwarehouse /var/borgwarehouse/[[\:xdigit\:]]*
borgwarehouse ALL=(ALL) NOPASSWD: /usr/bin/tee /var/borgwarehouse/[[\:xdigit\:]]*/.ssh/authorized_keys
borgwarehouse ALL=(ALL) NOPASSWD: /usr/bin/sed -ri s|*|g /var/borgwarehouse/[[\:xdigit\:]]*/.ssh/authorized_keys
```

This is the best compromise to improve safety and ensure high compatibility ?

**Comment by p-rintz** [comment]
Yes. It should make it (less likely, at least) so that somebody cant use the sudo rights for something outside what you need to do in borgwarehouse.

And it should work on basically any Linux that's still got any kind of support. Including Centos7. 
I cant really find out when wildcards were introduced in sudo, but I am finding blogs as far back as 2012 that talk about it. So at least then.

**Comment by Ravinou** [comment]
I have a permission denied on `getStorageUsed.sh`.

It's because we remove sudo in this script.

Even if we add users in borgwarehouse group, when I launch a backup with borgbackup, there is no rights on created files for group. 

Look : 
![image](https://user-images.githubusercontent.com/39600829/208249254-117c53ff-1686-4b23-b37c-d64e199e46ba.png)

So, I need to re-add sudo. I think this is the only script whith an error. What do you think ?

**Comment by p-rintz** [comment]
This is the case for repositories added before my PR.

I forgot to mention that it's a breaking change.

Old repositories do not have the correct permissions.

You will need to run the following for it to work:

`sudo chgrp -R borgwarehouse /var/borgwarehouse/
sudo chmod -R 750 /var/borgwarehouse
`

**Comment by Ravinou** [comment]
> This is the case for repositories added before my PR.

Unfortunately not. I can confirm you that once you have made a first backup with a borgbackup client, the `getStorageUsed.sh` script does not work without sudo. Because the files created by the client have no authorization at the group level, as on my image above. I just test it again with a fresh install.

Can you confirm me that it works for you after a first backup with a borgbackup client ? Thanks

**Comment by p-rintz** [comment]
It's possible the files get created with the wrong permissions afterward again. I'll have to test it later today.

**Comment by p-rintz** [comment]
Its what I thought. The permissions get reset to the user. But its also a simple fix. 

Following two changes:

This needs to be added to the createRepo.sh script. (I opened a  PR for that)
`sudo chmod -R g+s ${home}`

The chmod line in the sudoers file needs to be expanded to allow for the command to be run:
```
borgwarehouse ALL=(ALL) NOPASSWD: /usr/bin/chmod -R 750 /var/borgwarehouse/[[\:xdigit\:]]*, /usr/bin/chmod 600 /var/borgwarehouse/[[\:xdigit\:]]*/.ssh/authorized_keys, /usr/bin/chmod g+s /var/borgwarehouse/[[\:xdigit\:]]*
```

**Comment by p-rintz** [comment]
So. The issue is that the option I was trying to rely on what only changed in this PR on the borg side:
https://github.com/borgbackup/borg/pull/5226

Its only part of borg 1.2 and up from what I can see.

So there are two solutions for this then:
- Support borg 1.2.0 and up
- Re-add the root permissions to any read-only action

Considering that borg 2.0 is about to be released and that it would be the cleaner option for a server-side software such as borgwarehouse, Id vote for the first option.

Options for the documentation would be:
- Install the pip version
- Debian Stable has 1.2.2 in backports
- Next Debian stable has 1.2.2
- Ubuntu 22.04 has 1.2.0

Id go with the pip version, as that way any distribution would have the current borg version.

In the end, its your choice.

**Comment by Ravinou** [comment]
This is a big choice because it's a breaking change and it forces the user to use debian backports. Having a lot of experience on debian I never had any problems with backports but I know that it can bother many people who prefer to ensure maximum stability. 

And at the same time, switching to backports **simply allows to avoid sudo only on the storage script**, is that what we're talking about. Since it is only this script that is causing problems, I would prefer to adding the command in sudoers : 
```
borgwarehouse ALL=(ALL) NOPASSWD: /usr/bin/jc du -s *
```

And going back to the original file for `getStorageUsed.sh` :
```
cd /var/borgwarehouse
sudo jc du -s *
```

I know that's not perfect, but it avoids the use of backports, and I don't think that the return of `du` the is really a danger. What do you think about it ?

**Comment by p-rintz** [comment]
I agree its a breaking change. But I also think that it would be a good reason to do so.

Its not only the 'jc' command.

Its also the [helpers/shells/getLastSave.sh](https://github.com/Ravinou/borgwarehouse/blob/199fe64ae648d6b6695340bec2c7a20949c05426/helpers/shells/getLastSave.sh#L25) script.
And any future use of read-only commands you'd want/need to do by the application.

Backports is also not your only choice. And Id argue its the worse (default) choice.
I think moving to pip as the recommended way to install it (with ofc the option for people to use other sources, as long as the minimum required borg verson is met) is the best way forward, as it translates to other distributions as well.

It would also prime people to move to borg 2.0 in the future, which Id also start thinking about for support, as that is a breaking change from the side of borg and is coming out soon.
https://www.borgbackup.org/releases/borg-2.0.html

In general, I think going from the point of least-needed permissions is always best. Yes, its a breaking change, but it would solve this issue for any future occurrences where you need to access the repositories from the application.

**Comment by Ravinou** [comment]
I agree with your point of view. 

So, since it's a breaking change, I'm thinking of upgrading BorgWarehouse to v2.0 to incorporate these changes. Does that make sense to you too ?

But I think we have to anticipate three important things :
- updating the documentation (it won't take me long)
- create an upgrade procedure from v1.0 to v2.0 of BorgWarehouse.
- make sure that we don't forget any breaking changes related to this work on authorizations.
(- Do you see an other thing ?)

What do you think about it?

**Comment by p-rintz** [comment]
>  I'm thinking of upgrading BorgWarehouse to v2.0 to incorporate these changes. Does that make sense to you too ?

Sure. Sounds reasonable.

> make sure that we don't forget any breaking changes related to this work on authorizations.

Could you rephrase this? I am not entirely sure what you mean.

**Comment by Ravinou** [comment]
> Could you rephrase this? I am not entirely sure what you mean.

The theme of this issue is the improvement of sudo limitations. I just want to make sure that when we release a v2.0, we don't forget any major changes around this issue that would force us to release a v3.0 in a few weeks 🤪.

**Comment by p-rintz** [comment]
I cant really think of anything at the moment.
I can really only say anything about the shell script side, as I have no clue about javascript. 

If we get far enough into the future, where sudo version 1.9.10 is standard, and we could use regex based sudoers files, it wont really be a breaking change. 
Just a change in documentation.

What might be worthwhile to do is create some automated tests and look into how the docker side works more in detail. Especially with these changes. Maybe something might need to slightly change there.

I could look at the docker side sometime over the next few days.

**Comment by Ravinou** [comment]
> I can really only say anything about the shell script side, as I have no clue about javascript.
Don't worry, that's what it's about 👍

Docker is another subject, I've been working on it the last few days. I will open an issue to detail my analysis soon.

**Comment by p-rintz** [comment]
One thing we could possibly do is not release v2.0 right now, as well as have the non-optimal permissions in the v1.x branch for now.

Then make v2 of borgwarehouse target borg 2.0 and release it when that has released with any fixes/changes that might require.

**Comment by Ravinou** [comment]
I totally agree with this proposal. It is an excellent compromise.

**Comment by Ravinou** [comment]
I will update the documentation. To summarize, the file `/etc/sudoers.d/10-borgwarehouse` is : 

```
borgwarehouse ALL=(ALL) NOPASSWD: /usr/sbin/userdel -rf [[\:xdigit\:]]*
borgwarehouse ALL=(ALL) NOPASSWD: /usr/sbin/useradd -d /var/borgwarehouse/[[\:xdigit\:]]* -s /bin/bash -m [[\:xdigit\:]]*
borgwarehouse ALL=(ALL) NOPASSWD: /usr/bin/mkdir -p /var/borgwarehouse/[[\:xdigit\:]]*/repos/repo[[\:digit\:]]*, /usr/bin/mkdir -p /var/borgwarehouse/[[\:xdigit\:]]*/.ssh
borgwarehouse ALL=(ALL) NOPASSWD: /usr/bin/touch /var/borgwarehouse/[[\:xdigit\:]]*/.ssh/authorized_keys
borgwarehouse ALL=(ALL) NOPASSWD: /usr/bin/chmod -R 750 /var/borgwarehouse/[[\:xdigit\:]]*, /usr/bin/chmod 600 /var/borgwarehouse/[[\:xdigit\:]]*/.ssh/authorized_keys
borgwarehouse ALL=(ALL) NOPASSWD: /usr/bin/chown -R [[\:xdigit\:]]*\:borgwarehouse /var/borgwarehouse/[[\:xdigit\:]]*
borgwarehouse ALL=(ALL) NOPASSWD: /usr/bin/tee /var/borgwarehouse/[[\:xdigit\:]]*/.ssh/authorized_keys
borgwarehouse ALL=(ALL) NOPASSWD: /usr/bin/sed -ri s|*|g /var/borgwarehouse/[[\:xdigit\:]]*/.ssh/authorized_keys
borgwarehouse ALL=(ALL) NOPASSWD: /usr/bin/jc du -s [[\:xdigit\:]]*
borgwarehouse ALL=(ALL) NOPASSWD: /usr/bin/stat -c {"user"\:"%U"\,"lastSave"\:%Y} /var/borgwarehouse/[[\:xdigit\:]]*/repos/repo[[\:digit\:]]*/integrity.[[\:digit\:]]*
```

Are you agree with this ?

Thanks

**Comment by p-rintz** [comment]
Yeah, that should be correct for now.

**Comment by Ravinou** [comment]
The `--badname` was missing. I added it in documentation : 

```
borgwarehouse ALL=(ALL) NOPASSWD: /usr/sbin/userdel -rf [[\:xdigit\:]]*
borgwarehouse ALL=(ALL) NOPASSWD: /usr/sbin/useradd -d /var/borgwarehouse/[[\:xdigit\:]]* -s /bin/bash -m --badname [[\:xdigit\:]]*
borgwarehouse ALL=(ALL) NOPASSWD: /usr/bin/mkdir -p /var/borgwarehouse/[[\:xdigit\:]]*/repos/repo[[\:digit\:]]*, /usr/bin/mkdir -p /var/borgwarehouse/[[\:xdigit\:]]*/.ssh
borgwarehouse ALL=(ALL) NOPASSWD: /usr/bin/touch /var/borgwarehouse/[[\:xdigit\:]]*/.ssh/authorized_keys
borgwarehouse ALL=(ALL) NOPASSWD: /usr/bin/chmod -R 750 /var/borgwarehouse/[[\:xdigit\:]]*, /usr/bin/chmod 600 /var/borgwarehouse/[[\:xdigit\:]]*/.ssh/authorized_keys
borgwarehouse ALL=(ALL) NOPASSWD: /usr/bin/chown -R [[\:xdigit\:]]*\:borgwarehouse /var/borgwarehouse/[[\:xdigit\:]]*
borgwarehouse ALL=(ALL) NOPASSWD: /usr/bin/tee /var/borgwarehouse/[[\:xdigit\:]]*/.ssh/authorized_keys
borgwarehouse ALL=(ALL) NOPASSWD: /usr/bin/sed -ri s|*|g /var/borgwarehouse/[[\:xdigit\:]]*/.ssh/authorized_keys
borgwarehouse ALL=(ALL) NOPASSWD: /usr/bin/jc du -s [[\:xdigit\:]]*
borgwarehouse ALL=(ALL) NOPASSWD: /usr/bin/stat -c {"user"\:"%U"\,"lastSave"\:%Y} /var/borgwarehouse/[[\:xdigit\:]]*/repos/repo[[\:digit\:]]*/integrity.[[\:digit\:]]*
```

It's due to my Fedora improvement commit : https://github.com/Ravinou/borgwarehouse/commit/199fe64ae648d6b6695340bec2c7a20949c05426

Now there is still a problem with the checkStatus script that triggers a sudo password request. I'll look at it later.

**Comment by p-rintz** [comment]
> Now there is still a problem with the checkStatus script that triggers a sudo password request. I'll look at it later.

That is strange. It worked on my side.

With auditd + `ausearch -ui borgwarehouse -i | grep failed` you can get more information on the failed sudo commands of the borgwarehouse user.

**Comment by Ravinou** [comment]
The problem is present if :
- if no repository is created
- if there is only one repository and there has never been a backup on it

Do you have the same case @p-rintz ?

**Comment by Ravinou** [comment]
For me this is a normal error and not dependent on `sudo.` The `integrity` file does not exist until a first backup has been made. However, the "cannot stat" error can be ignored and is not blocking. But with `sudo`, it becomes blocking for the application and it will launch background processes that will be waiting for the `sudo` password.

I think we have to change the logic of the script rather than looking for a solution with `sudo`.

**Comment by Ravinou** [comment]
Ok, sudo is actually not needed at all on this script... as simple as that

**Comment by p-rintz** [comment]
> Ok, sudo is actually not needed at all on this script... as simple as that

It should be as soon as somebody makes a backup. At least for version 1.0 where the umask isnt working.

> type=USER_CMD msg=audit(12/22/2022 12:46:45.246:126) : pid=741498 uid=borgwarehouse auid=root ses=63159 subj==unconfined msg='cwd=/home/borgwarehouse/borgwarehouse cmd=/usr/bin/stat -c {user:%U,lastSave:%Y} /var/borgwarehouse/*/repos/repo*/integrity.* exe=/usr/bin/sudo terminal=pts/1 res=failed'

This is strange to me. Because on my side the globbing gets split down into its parts like this:

**Comment by Ravinou** [comment]
I have the same error on getStorage if there is no repository. Sudo is useless too on this script, i'll remove it.

**Comment by p-rintz** [comment]
As soon as a user creates a backup, the permissions will be changed to {$user}:{$user} and then sudo will be needed. At least in the 1.0 branch.

**Comment by Ravinou** [comment]
> As soon as a user creates a backup, the permissions will be changed to {$user}:{$user} and then sudo will be needed. At least in the 1.0 branch.

It's correct for `getStorageUsed.sh`. 

But for `getLastSave.sh` i never need sudo. Moreover, sudo is a problem if the `integrity` file is not present like I said before. So I will remove it from the script.: 
![image](https://user-images.githubusercontent.com/39600829/209155038-cd64e543-df9b-4bf5-8cba-53e4e2a9814f.png)

**Comment by p-rintz** [comment]
Im confused. This is on my side, I specifically changed the group back to the user only:

borgwarehouse@no1-ter:/root$ ls -lah /var/borgwarehouse/026b1c3b/repos/repo1/
total 11M
drwxr-s--- 3 026b1c3b borgwarehouse 4.0K Dec 14 08:00 .
drwxr-s--- 3 026b1c3b borgwarehouse 4.0K Dec 12 12:50 ..
-rwxr-s--- 1 026b1c3b borgwarehouse  975 Dec 12 12:50 config
drwxr-s--- 3 026b1c3b borgwarehouse 4.0K Dec 12 12:50 data
-rwxr-s--- 1 026b1c3b borgwarehouse  104 Dec 14 08:00 hints.22
-rwxr-s--- 1 026b1c3b borgwarehouse  11M Dec 14 08:00 index.22
-rwxr-x--- 1 026b1c3b 026b1c3b       190 Dec 14 08:00 integrity.22
-rwxr-s--- 1 026b1c3b borgwarehouse   16 Dec 14 08:00 nonce
-rwxr-s--- 1 026b1c3b borgwarehouse   73 Dec 12 12:50 README
borgwarehouse@no1-ter:/root$ stat -c {"user":"%U","lastSave":%Y} /var/borgwarehouse/026b1c3b/repos/repo1/integrity.22
stat: cannot statx 'lastSave:%Y': Permission denied

**Comment by Ravinou** [comment]
😅 WTF, Are we going crazy? Are you running debian 11.5 like me?

**Comment by p-rintz** [comment]
```
root@no1-ter:~# cat /etc/debian_version 
11.6
```
Your user has no read permissions to the file. I have no idea why It is working?!

**Comment by Ravinou** [comment]
I fixed the execution of shell for getStorage directly in API. It's okay now if repo list is empty.

**Comment by Ravinou** [comment]
So, final release for sudoers file is : 

```
borgwarehouse ALL=(ALL) NOPASSWD: /usr/sbin/userdel -rf [[\:xdigit\:]]*
borgwarehouse ALL=(ALL) NOPASSWD: /usr/sbin/useradd -d /var/borgwarehouse/[[\:xdigit\:]]* -s /bin/bash -m --badname [[\:xdigit\:]]*
borgwarehouse ALL=(ALL) NOPASSWD: /usr/bin/mkdir -p /var/borgwarehouse/[[\:xdigit\:]]*/repos/repo[[\:digit\:]]*, /usr/bin/mkdir -p /var/borgwarehouse/[[\:xdigit\:]]*/.ssh
borgwarehouse ALL=(ALL) NOPASSWD: /usr/bin/touch /var/borgwarehouse/[[\:xdigit\:]]*/.ssh/authorized_keys
borgwarehouse ALL=(ALL) NOPASSWD: /usr/bin/chmod -R 750 /var/borgwarehouse/[[\:xdigit\:]]*, /usr/bin/chmod 600 /var/borgwarehouse/[[\:xdigit\:]]*/.ssh/authorized_keys
borgwarehouse ALL=(ALL) NOPASSWD: /usr/bin/chown -R [[\:xdigit\:]]*\:borgwarehouse /var/borgwarehouse/[[\:xdigit\:]]*
borgwarehouse ALL=(ALL) NOPASSWD: /usr/bin/tee /var/borgwarehouse/[[\:xdigit\:]]*/.ssh/authorized_keys
borgwarehouse ALL=(ALL) NOPASSWD: /usr/bin/sed -ri s|*|g /var/borgwarehouse/[[\:xdigit\:]]*/.ssh/authorized_keys
borgwarehouse ALL=(ALL) NOPASSWD: /usr/bin/jc du -s [[\:xdigit\:]]*
```

The documentation of borgwarehouse is up to date.

I hope that we have now done the trick for this version 1.0 😅

We'll do the rest when BorgBackup 2.0 is released and when we switch to BorgWarehouse 2.0 to support this version as we mentioned before

**Comment by Ravinou** [comment]
This issue is closed. Sudo permission is now no longer required at all in v2.0 : https://github.com/Ravinou/borgwarehouse/releases/tag/v2.0.0

Thanks again @p-rintz for your help here 🙏
